### PR TITLE
Implement error applet

### PR DESCRIPTION
**Pull request by Pirater12**
This would be useful to show custom error messages!

### CODE REVIEW

File: libctru/include/3ds/applets/error.h [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+
+    bool        softwareReset;
+
+    bool        appJump;
+
+    ReturnCode  returnCode;
+
+    u16         eulaVersion;
+
+}ErrConf;
```
Comment (mtheall) [inline libctru/include/3ds/applets/error.h]: No need to double-space

File: libctru/source/applets/error.c [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+	ssize_t units = utf8_to_utf16(out, (const uint8_t*)in, max);
+	if (units < 0)
+	{
+		out[0] = 0;
+		return;
+	}
+
+	out[units] = 0;
+}
+char k[1900];
```
Comment (mtheall) [inline libctru/source/applets/error.c]: Should be static.

A size of `1900` can be too restrictive. For example, you can have a UTF8 string longer than 1900 bytes that gets encoded to less than 1900 UTF16 code points. The limit of `1900` would prevent such messages from being usable when the UTF16 encoding supports them.

In the end, I would probably ditch this for the method I describe below.

File: libctru/source/applets/error.c [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+char *c_shift(char text[])
```
Comment (mtheall) [inline libctru/source/applets/error.c]: Should be static. I'm not really sure what is the purpose of shifting the buffer by one and duplicating the first byte, but this would be equivalent and more readable:

```
static char* c_shift(const char *text)
{
	size_t len = strlen(text);
	char *str = (char*)malloc(len + 2);
	if(str)
	{
		str[0] = text[0];
		strcpy(&str[1], text);
	}
	return str;
}
```

File: libctru/source/applets/error.c [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+{ 
+  int i=strlen(text);
+  char tex[i];
```
Comment (mtheall) [inline libctru/source/applets/error.c]: I would avoid allocating a buffer of unbounded size on the stack

File: libctru/source/applets/error.c [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+  tex[0]=text[0];
+  int j=1;
+  for(int a=0;a<=i;a++)
+  {
+	  tex[j]=text[a];
+	  j++;
+  }
+	strncpy(k, tex, i+1);
```
Comment (mtheall) [inline libctru/source/applets/error.c]: I think you mean to use `1900` instead of `i+1`

File: libctru/source/applets/error.c [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+	return k;
+}
+void error_code(ErrConf* err,int error)
```
Comment (mtheall) [inline libctru/source/applets/error.c]: Empty line before this function please

File: libctru/source/applets/error.c [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+	err->useLanguage=       lang;
+	err->upperScreenFlag= NORMAL;
+	err->eulaVersion =         0;
+	err->homeButton  =      true;
+        err->softwareReset =   false;
+        err->appJump     =     false;
+        err-> returnCode   = UNKNOWN;
+}
+
+static void errorConvertToUTF16(u16* out, const char* in, int max)
```
Comment (mtheall) [inline libctru/source/applets/error.c]: Use `size_t max`

File: libctru/source/applets/error.c [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+{
+	if (!in || !*in)
+	{
+		out[0] = 0;
+		return;
+	}
+
+	ssize_t units = utf8_to_utf16(out, (const uint8_t*)in, max);
```
Comment (mtheall) [inline libctru/source/applets/error.c]: Use `max-1`, otherwise you can get `max` returned which will cause you to write the nul-terminator to `out[max]` which is out-of-bounds.

File: libctru/source/applets/error.c [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+}
+void error_code(ErrConf* err,int error)
+{
+	err->errorCode = error;
+}
+
+void error_text(ErrConf *err, char* text)
+{   
+    char *tex=c_shift(text);
+    errorConvertToUTF16(err->Text, tex,1900);
```
Comment (mtheall) [inline libctru/source/applets/error.c]: Put a space before `1900`.
`free(tex)` if using my suggested `c_shift()` function.

File: libctru/source/applets/error.c [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+}
+
+void error_disp(ErrConf* err)
+{   
+	 aptLaunchLibraryApplet(APPID_ERROR , err, sizeof(*err), 0);
```
Comment (mtheall) [inline libctru/source/applets/error.c]: Remove the space after `APPID_ERROR`

**Comment by Pirater12** [comment]
I have fixed the files according to the review,please check it out.

File: libctru/include/3ds/applets/error.h [hunk 11]
```diff
@@ -0,0 +1,112 @@
+/**
+ * @file error.h
+ * @brief Error applet.
+ */
+#pragma once
+
+ enum
```
Comment (deltabeard) [inline libctru/include/3ds/applets/error.h]: Remove space before enum

File: libctru/source/applets/error.c [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+	return str;
+}
+
+void error_code(ErrConf* err,int error)
+{
+	err->errorCode = error;
+}
+
+void error_text(ErrConf *err, char* text)
+{	char *tex=c_shift(text);
```
Comment (deltabeard) [inline libctru/source/applets/error.c]: space before and after '=' to stay consistent with your code style.

File: libctru/source/applets/error.c [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+	}
+
+	out[units] = 0;
+}
+
+static char* c_shift(const char *text)
+{
+	size_t len = strlen(text);
+	char *str = (char*)malloc(len + 2);
+	if(str)
```
Comment (deltabeard) [inline libctru/source/applets/error.c]: Space after 'if' to stay consistent with your code style.

File: libctru/include/3ds/applets/error.h [hunk 14]
```diff
@@ -0,0 +1,112 @@
...
+}ErrorType;
+
+
+///< Flags for the Upper Screen.Does nothing even if specified.
+
+typedef enum
+{
+	NORMAL = 0,    
+	STEREO,        
+}screen_flag;
```
Comment (deltabeard) [inline libctru/include/3ds/applets/error.h]: You switch between using camel case and underscore variable names throughout.

**Comment by Pirater12** [comment]
Ok, Done !!

**Comment by fuzun** [comment]
Stop doing stupid updates. My inbox just got fucked up because of this mess...

File: libctru/include/3ds/applets/error.h [hunk 15]
```diff
@@ -0,0 +1,111 @@
...
+	GERMAN,        
+	ITALIAN,       
+	SPANISH,       
+	SIMP_CHINESE,  
+	KOREAN,        
+	DUTCH,         
+	PORTUGUESE,    
+	RUSSIAN,       
+	TRAD_CHINESE,  
+}ErrorLang;
```
Comment (fincs) [inline libctru/include/3ds/applets/error.h]: This looks a lot like [CFG_Language](https://github.com/smealum/ctrulib/blob/master/libctru/include/3ds/services/cfgu.h#L19).

File: libctru/include/3ds/applets/error.h [hunk 16]
```diff
@@ -0,0 +1,111 @@
...
+}ErrorType;
+
+
+///< Flags for the Upper Screen.Does nothing even if specified.
+
+typedef enum
+{
+	NORMAL = 0,    
+	STEREO,        
+}ScreenFlag;
```
Comment (fincs) [inline libctru/include/3ds/applets/error.h]: These unprefixed identifiers name pollute the global namespace. Can you prefix the enums with something like `ERROR_`? Also, ScreenFlag -> errorScreenFlag

File: libctru/include/3ds/applets/error.h [hunk 17]
```diff
@@ -0,0 +1,111 @@
...
+	EULA,                                                       ///< Displays the EULA
+	ERROR_TYPE_EULA_FIRST_BOOT,                                 ///< Use prohibited.
+	ERROR_TYPE_EULA_DRAW_ONLY,                                  ///< Use prohibited.
+	ERROR_TYPE_AGREE,                                           ///< Use prohibited.
+	ERROR_CODE_LANGUAGE = ERROR_CODE | ERROR_LANGUAGE_FLAG,     ///< Displays a network error message in a specified language.
+	ERROR_TEXT_LANGUAGE = ERROR_TEXT | ERROR_LANGUAGE_FLAG,     ///< Displays text passed to this applet in a specified language.
+	EULA_LANGUAGE = EULA | ERROR_LANGUAGE_FLAG,                 ///< Displays EULA in a specified language.
+	ERROR_TEXT_WORD_WRAP = ERROR_TEXT | ERROR_WORD_WRAP_FLAG,   ///< Displays the custom error message passed to this applet with automatic line wrapping
+	ERROR_TEXT_LANGUAGE_WORD_WRAP = ERROR_TEXT | ERROR_LANGUAGE_FLAG | ERROR_WORD_WRAP_FLAG,   ///< Displays the custom error message with automatic line wrapping and in the specified language.
+}ErrorType;
```
Comment (fincs) [inline libctru/include/3ds/applets/error.h]: ErrorType -> errorType

File: libctru/include/3ds/applets/error.h [hunk 18]
```diff
@@ -0,0 +1,111 @@
...
+typedef enum 
+{
+	UNKNOWN = -1,       
+	NONE    = 0,        
+	SUCCESS,            
+	NOT_SUPPORTED,      
+	HOME_BUTTON = 10,   
+	SOFTWARE_RESET,     
+	POWER_BUTTON,       
+}ReturnCode;
```
Comment (fincs) [inline libctru/include/3ds/applets/error.h]: -> errorReturnCode

File: libctru/include/3ds/applets/error.h [hunk 19]
```diff
@@ -0,0 +1,111 @@
...
+	int errorCode;
+	ScreenFlag upperScreenFlag;
+	ErrorLang useLanguage;
+	u16 Text[1900];
+	bool homeButton;
+	bool softwareReset;
+	bool appJump;
+	ReturnCode returnCode;
+	u16 eulaVersion;
+}ErrConf;
```
Comment (fincs) [inline libctru/include/3ds/applets/error.h]: -> errorConf

File: libctru/include/3ds/applets/error.h [hunk 20]
```diff
@@ -0,0 +1,111 @@
...
+/**
+* @brief Init the error applet.
+* @param err Pointer to ErrConf.
+* @param type ErrorType Type of error.
+* @param lang ErrorLang Lang of error. 
+*/
+void error_Init(ErrConf* err, ErrorType type, ErrorLang lang);
```
Comment (fincs) [inline libctru/include/3ds/applets/error.h]: Remove underscore in function names (errorInit, errorCode, errorText, errorDisp)

File: libctru/include/3ds/applets/error.h [hunk 21]
```diff
@@ -0,0 +1,111 @@
...
+* @param err Pointer to ErrConf.
+* @param error Error-code to display.
+*/
+void error_code(ErrConf* err, int error);
+/**
+* @brief Sets error text to display.
+* @param err Pointer to ErrConf.
+* @param text Error-text to display.
+*/
+void error_text(ErrConf* err, char* text);
```
Comment (fincs) [inline libctru/include/3ds/applets/error.h]: Should be const char* text

File: libctru/source/applets/error.c [hunk 22]
```diff
@@ -0,0 +1,69 @@
...
+#include <string.h>
+#include <3ds/types.h>
+#include <3ds/svc.h>
+#include <3ds/synchronization.h>
+#include <3ds/services/apt.h>
+#include <3ds/services/cfgu.h>
+#include <3ds/util/utf.h>
+#include <3ds/applets/error.h>
+
+void error_Init(errorConf* err, errorType type, CFG_Language lang)
```
Comment (fincs) [inline libctru/source/applets/error.c]: Still forgot to rename here.

**Comment by Pirater12** [comment]
@fincs Fixed.

**Comment by Pirater12** [comment]
Um, any updates?

**Comment by WinterMute** [comment]
shhh, it's Switch launch day. Everyone is playing Zelda :p

**Comment by deltabeard** [comment]
Bumping this to regain the attention of the reviewers.

**Comment by Pirater12** [comment]
Bump..

File: libctru/source/applets/error.c [hunk 23]
```diff
@@ -0,0 +1,69 @@
...
+	if (units < 0)
+	{
+		out[0] = 0;
+		return;
+	}
+
+	out[units] = 0;
+}
+
+static char* c_shift(const char *text)
```
Comment (fincs) [inline libctru/source/applets/error.c]: Why is this needed?

Comment (Pirater12) [inline libctru/source/applets/error.c]: `c_shift()` is needed because it seems that after using function `utf8toutf16` the first character of the input string is not present in the output .Not sure why this is happening.”Hello World” would be displayed as “ello World” in error Applet without using c_shift().

Comment (mtheall) [inline libctru/source/applets/error.c]: Seems likely that Text might be misaligned. Can you try adding a padding u16 before Text, which probably means you also need to shorten it by 1 also.

Comment (mtheall) [inline libctru/source/applets/error.c]: Actually maybe it's the opposite. Is Text one u16 too early? Maybe one of those enums is supposed to be 16 bits wide. I'm pretty sure as-is they are each 32-bit.

Comment (Pirater12) [inline libctru/source/applets/error.c]: Thanks @mtheall. `c_shift` is no longer needed!
